src/net/tcp/nbd.c

```diff
@@ -0,0 +1,1095 @@
+ * @v len		Length of data buffer
+ * @ret rc		Return status code
+
+ */
+static int nbd_block_read ( struct nbd_session *nbd,
+			    struct interface *block,
+			    uint64_t lba, unsigned int count,
+			    userptr_t buffer, size_t len ) {
+	int rc;
+
+	DBGCP ( nbd, "NBD %p block %p read LBA 0x%08llx count 0x%04x\n",
+		nbd, block, lba, count );
+
+	rc = nbd_command ( nbd, block, NBD_BLOCK_READ,
+			   lba, count, buffer, len );
+	if ( rc != 0 )
+		return rc;
+
+	/* Start command processing */
+	nbd_start_tx ( nbd, NBD_TX_CMD_HEADER );
+	return 0;
+}
+
+/**
+ * Issue NBD block write
+ *
+ * @v nbd		NBD session
+ * @v block		Block data interface
+ * @v lba		Starting logical block address
+ * @v count		Number of blocks to transfer
+ * @v buffer		Data buffer
+ * @v len		Length of data buffer
+ * @ret rc		Return status code
+ */
+static int nbd_block_write ( struct nbd_session *nbd,
+			     struct interface *block,
+			     uint64_t lba, unsigned int count,
+			     userptr_t buffer, size_t len ) {
+	int rc;
+
+	DBGCP ( nbd, "NBD %p block %p write LBA 0x%08llx count 0x%04x\n",
+		nbd, block, lba, count );
+
+	if ( nbd->trans_flags & NBD_FLAG_READ_ONLY ) {
+		DBGC ( nbd, "NBD %p read only media\n", nbd );
+		return -EROFS;
+	}
+
+	rc = nbd_command ( nbd, block, NBD_BLOCK_WRITE,
+			   lba, count, buffer, len );
+	if ( rc != 0 )
+		return rc;
+
+	/* Start command processing */
+	nbd_start_tx ( nbd, NBD_TX_CMD_HEADER );
+	return 0;
+}
+
+/**
+ * Read NBD device capacity
+ *
+ * @v nbd		NBD session
+ * @v block		Block data interface
+ * @ret rc		Return status code
+ */
+static int nbd_block_read_capacity ( struct nbd_session *nbd,
+				     struct interface *block ) {
+	int rc;
+
+	DBGCP ( nbd, "NBD %p block read capacity %p\n", nbd, block );
+
+	rc = nbd_command ( nbd, block, NBD_BLOCK_READ_CAPACITY,
+			   0, 0, UNULL, 0 );
+	if ( rc != 0 )
+		return rc;
+
+	/* Start command processing */
+	nbd_start_tx ( nbd, NBD_TX_CMD_BRC );
+	return 0;
+}
+
+/**
+ * Describe as an EFI device path
+ *
+ * @v nbd		NBD session
+ * @ret path		EFI device path, or NULL on error
+ */
+static EFI_DEVICE_PATH_PROTOCOL * nbd_efi_describe ( struct nbd_session *nbd ) {
+	DBGCP ( nbd, "NBD %p EFI describe\n", nbd );
+	return efi_uri_path ( nbd->uri );
+}
+
+
+/** NBD block interface operations */
+static struct interface_operation nbd_block_op[] = {
+	INTF_OP ( xfer_window, struct nbd_session *, nbd_block_window ),
+	INTF_OP ( block_read, struct nbd_session *, nbd_block_read ),
+	INTF_OP ( block_write, struct nbd_session *, nbd_block_write ),
+	INTF_OP ( block_read_capacity, struct nbd_session *,
+		  nbd_block_read_capacity ),
+	INTF_OP ( intf_close, struct nbd_session *, nbd_close ),
+	EFI_INTF_OP ( efi_describe, struct nbd_session *, nbd_efi_describe ),
+};
+
+/** NBD block interface descriptor */
+static struct interface_descriptor nbd_block_desc =
+	INTF_DESC ( struct nbd_session, block, nbd_block_op );
+
+
+static inline void nbd_neg_done ( struct nbd_session *nbd ) {
```

Comment:
Please give this function a comment block to explain what it is for

Reply:
OK, comment added, function moved near relevant functions